src/capped_hashmap.rs

```diff
@@ -0,0 +1,172 @@
+use std::{
+    cmp::Eq,
+    collections::{HashMap, VecDeque},
+    hash::Hash,
+};
+
+pub struct CappedHashMap<K, V>
+where
+    K: Hash + Eq + Copy + Clone,
+{
+    capacity: usize,
+    inner: HashMap<K, V>,
+    last_items: VecDeque<K>,
+}
+
+impl<K, V> CappedHashMap<K, V>
+where
+    K: Hash + Eq + Copy + Clone,
+{
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            // We add a little bit margin to the capacity, because we always insert first and then check
+            // if collection is full. So we don't want the hashmap to double it's capacity because of that.
+            capacity: capacity + 1,
+            inner: HashMap::with_capacity(capacity),
+            last_items: VecDeque::with_capacity(capacity),
+        }
+    }
+
+    /// Inserts an new item to the collection. Return Some(key) where key is the
+    /// key that was removed when we reach the max capacity. Otherwise returns None.
+    pub fn insert(&mut self, k: K, v: V) -> Option<K> {
+        let mut ret = None;
+
+        // replacing a value should not push any new items to last_items
+        if let None = self.inner.insert(k, v) {
+            self.last_items.push_front(k);
+        }
+
+        if self.last_items.len() > self.capacity - 1 {
+            // remove the oldest item. We an safely unwrap because we know the last_items is not empty at this point
+            let key = self.last_items.pop_back().unwrap();
+            assert!(self.remove(&key).is_some());
+
+            ret = Some(key);
+        }
+
+        ret
+    }
+
+    /// Removes a key from the map, returning the value at the key if the key was previously in the map.
+    pub fn remove(&mut self, k: &K) -> Option<V> {
+        let Some(v) = self.inner.remove(k) else {
+            return None;
+        };
+
+        self.last_items = self
+            .last_items
+            .iter()
+            .filter(|key| *key != k)
+            .map(|key| *key)
+            .collect::<VecDeque<_>>();
```

Comment:
it kinda sucks that we have to go through the whole thing here :D I agree that we should assume that the key to remove is one of the last one appended (and that the oldest stuff is probably just stale stuff at this point). Maybe a LinkedList is better as removing something at any index is easier? (so find followed by a remove). In any case I think this would be better:

```rust
self.last_items.iter().position(|key| key == &k).and_then(self.last_items.remove(pos));
```

Reply:
a LinkedList isn't faster on [removals](https://doc.rust-lang.org/std/collections/struct.LinkedList.html#method.remove). 

> This operation should compute in O(n) time.

Reply:
ah yeah you'd need some sort of linkedlist + hashmap to know which two nodes to update to remove a node :D